scripts/testCardTypes.js

```diff
@@ -1,2 +1,137 @@
-// @todo: Put valueable logic here
-console.log('This file should check files structure of card-types.');
+/* eslint-disable global-require,import/no-dynamic-require */
+const { lstatSync, readdirSync, writeFileSync } = require('fs');
+const { join } = require('path');
+const { tmpdir } = require('os');
+const AnkiExport = require('anki-apkg-export').default;
+
+const OUTPUT_DIR = tmpdir();
+
+require('json5/lib/register');
+
+const isDirectory = source => lstatSync(source).isDirectory();
+const getDirectories = source =>
+  readdirSync(source)
+    .map(name => join(source, name))
+    .filter(isDirectory);
+
+const currentTypes = getDirectories('./types');
+
+const testErrors = {};
+
+const containsFile = fileName => typeDir => {
+  let ret = true;
+  try {
+    lstatSync(`./${typeDir}/${fileName}`);
+  } catch (e) {
+    console.log(e.message);
+    ret = false;
+  }
+  return ret;
+};
+
+const containsReadme = containsFile('README.md');
+const containsExample = containsFile('example.json5');
+const containsValidator = containsFile('validator.js');
+const containsIndex = containsFile('index.js');
+
+const testValidator = typeDir => {
+  let ret = true;
+  try {
+    const failedObj = {};
+    const successObj = require(`../${typeDir}/example.json5`);
+    const validator = require(`../${typeDir}/validator.js`);
+    if (!validator(successObj)) {
+      throw new Error(`Validator from ${typeDir} should pass example object`);
+    }
+    if (validator(failedObj)) {
+      throw new Error(`Validator from ${typeDir} should fail dummy object`);
+    }
+  } catch (e) {
+    console.log(`testValidator error:`, e);
+    ret = false;
+  }
+  return ret;
+};
+
+const testIndex = typeDir => {
+  let ret = true;
+  try {
+    const successObj = require(`../${typeDir}/example.json5`);
+    const index = require(`../${typeDir}/index.js`);
+    const result = index(successObj);
+    if (typeof result.front !== 'string') {
+      throw new Error('should have `front property`');
+    }
+    if (typeof result.back !== 'string') {
+      throw new Error('should have `back` property');
+    }
+    if (!Array.isArray(result.tags)) {
+      throw new Error('should have `tags` property');
+    }
+  } catch (e) {
+    console.log(`testValidator error:`, e);
+    ret = false;
+  }
+  return ret;
+};
+
+const testAnkiCardCreation = async typeDir => {
+  let ret = true;
+  try {
+    const apkg = new AnkiExport(`${typeDir} example deck`);
+    const index = require(`../${typeDir}/index.js`);
+    const example = require(`../${typeDir}/example.json5`);
+    const { front, back, tags } = index(example);
+    apkg.addCard(front, back, tags);
+
+    await apkg
+      .save()
+      .then(zip =>
+        writeFileSync(
+          `${OUTPUT_DIR}/${typeDir.replace(/\W/g, '_')}.apkg`,
```

Comment:
I would like to know, why do you use replace here? I think I don't catch the idea of doing that.

Reply:
`typeDir` contains `types/type_name` - you can't create file in subdirectory without creation dir before that, so I remove `/` symbol to avoid this situation